### PR TITLE
Gp/continuous geometry

### PR DESCRIPTION
**Pull request by gpescia**
First draft to implement the geometry for continuous space. 
There is continous_geometry.py as base class for specialized ones: Cell, Free. 
At the moment only these two are available (so no mix of boundary conditions). 
I commented out all tests in hilbert that test for mixed boundary conditions as well as the Langevin sampler. 
Examples run fine on my machine.

### CODE REVIEW

File: netket/hilbert/continuous_hilbert.py [hunk 1]
```diff
@@ -26,42 +26,17 @@ class ContinuousHilbert(AbstractHilbert):
...
-                """`pbc` must be either a bool or a tuple indicating the periodicity of each spatial dimension."""
-            )
-
-        if np.any(np.logical_and(np.isinf(self._extent), pbc)):
-            raise ValueError(
-                "If you do have periodic boundary conditions in a given direction the maximum of the quantum number "
-                "in that direction must be finite."
-            )
-
+        self._pbc = geometry.pbc
```
Comment (gcarleo) [inline netket/hilbert/continuous_hilbert.py:37]: I guess this could be removed if we use distances defined in geometry

**Comment by PhilipVinc** [comment]
- We should decide where to put the geometry stuff. I'm not really in favour of `.graph` let's see what @gcarleo thinks about
 - We should not break old interfaces. So ContinuousHilbert should still accept the old arguments, and also `geometry`. If geometry is none it should use the other arguments to build a geomtry
 - This way of implementing the random states is not ok

Before actually implementing this stuff, I think someone should lay out clearly what we want to do with it. For example, do we need the inverse lattice? Why? What is the expected use case of this?

For example, the messy translation code in the sampling makes me suspect that the abstract class should have a method `translate_positions_by_vector` or something like that...

**Comment by gpescia** [comment]
> We should not break old interfaces. So ContinuousHilbert should still accept the old arguments, and also geometry. If geometry is none it should use the other arguments to build a geomtry

yes true, I‘ll add the other as well.

> This way of implementing the random states is not ok

In the end I guess we want to dispatch on cell/free. 

> For example, do we need the inverse lattice? Why? What is the expected use case of this?

yes to go from lattice basis to standard basis and back.

**Comment by PhilipVinc** [comment]
> yes to go from lattice basis to standard basis and back.

In what part of the code is this needed? If this is not used in the code, can you give an example of what it can be used for and how it would be used?

**Comment by gcarleo** [comment]
Yes for the position of the module, it's a secondary problem, but I also wouldn't put it in Graph. Anyways, there is time to think about that... 

In my view the Geometry object should have a minimal interface to start with allowing to:

1. Compute distances between any two points R1,R2 that are valid for the given Geometry. This method is just `distance` and should know nothing about periodicity etc explicitly (i.e. you don't pass pbc=True etc to `distance') it is the essence of the Geometry to know how the distance should be computed (e.g. with minimum image etc)
2. Generate a valid random point R 
3. Add/subtract vectors in this space (say implement R_1 + R_2)? (not sure about this point but maybe it's handy for sampling)

**Comment by gpescia** [comment]
In the sampler and the hilbert space random state. We sample fractional coordinates, for this we need to go from standard to lattice basis back to standard

**Comment by gcarleo** [comment]
> In the sampler and the hilbert space random state. We sample fractional coordinates, for this we need to go from standard to lattice basis back to standard

but that's why the random points should be generated in the Geometry, not in Hilbert

**Comment by gpescia** [comment]
1) Well you need to tell the function whether you want periodic in the cell or minimum image convention. For this you need the ‚mode’ I would say. In the free case you don‘t need it.

2) The only difference is the placement of the random state right? One still goes nack and forth between the two bases.

**Comment by gcarleo** [comment]
1. Yes but you specify periodic or minimum image when you construct Cell, not when you call Distance, in this way you completely separate the sampling and Hilbert from these details of the geometry 

2. It's important to disentangle everything that relates to Geometry from Hilbert and the Sampling, so that you have a general code for the last two that can work for any Geometry

**Comment by PhilipVinc** [comment]
> Well you need to tell the function whether you want periodic in the cell or minimum image convention. For this you need the ‚mode’ I would say. In the free case you don‘t need it.

Are there cases where for the same geometry one wants one of those two different distances? Can you give an example? 
Can we think a case of a `default` distance to be used mostly everywhere? In that case `Geometry` objects could expose different distance functions, and a default one to be used for interaction potentials or I don't know what else.

**Comment by gcarleo** [comment]
I think there is no reasonable case in which you want to use two different distances within the same simulation Cell, that's why I think the Cell type should be used to identify uniquely the type of distance used as well

**Comment by gpescia** [comment]
Well potential energies are computed always with minimum image, and distances for the wf never. So you need both…

**Comment by PhilipVinc** [comment]
What do you mean by `distances for the wf` ?

**Comment by gpescia** [comment]
> It's important to disentangle everything that relates to Geometry from Hilbert and the Sampling, so that you have a general code for the last two that can work for any Geometry

But random state is actually a state in hilbert not in the geometry, right?

**Comment by gpescia** [comment]
distance features used in the wavefunction

**Comment by gcarleo** [comment]
sorry Gabriel but is it truly necessary to use two different distances? I would say one should be enough / is much nicer than using two different distances for the same space

**Comment by gcarleo** [comment]
in any case, maybe it's more efficient to talk on zoom tomorrow or so

**Comment by gpescia** [comment]
Sure we can have a meeting! I‘m just saying that there are two distance measures which both get used. I‘m not sure how to optimally resolve that..

**Comment by PhilipVinc** [comment]
I would love to understand why you are using more than one distance. 
If I understand it's because in the nn/variational ansatz you want to use the standard distance instead of the minimum image.
But physically this, I think, is wrong, because the minimum image is the physical one in a periodic system.
So what is the underlying reason of this? Performance? It works? Or maybe I fail to understand?

**Comment by gpescia** [comment]
The reason is smootheness. The minimum image convention reproduces Euclidean distance as you say but it is not smooth and cannot be used as input to the wavefunction as we take derivatives w.r.t. position. The other distance (sin-distance) behaves similar as the Euclidean distance but is smooth. So for potential energies (no derivatives involved) we want the "correct" euclidean distance. For the wavefunction we take the less correct but smooth version.

**Comment by PhilipVinc** [comment]
Thanks! This was the explanation I was looking for.
I think that to maintain flexibility we could should just implement both as two different functions. Then in the models, one uses what he wants.
But to be further discussed..

**Comment by gcarleo** [comment]
OK, not ideal, but then this is solved maybe specifying what type of distance you want, it's true that even in euclidean space you might want L1,L2... etc

**Comment by gpescia** [comment]
sure we can do that too. But that is basically the same as setting a 'mode' flag or do you gain performance with this?

**Comment by gcarleo** [comment]
Yes I think it's just a flag to be passed: `geo.Distance(R1,R2,type='L2')` etc

**Comment by PhilipVinc** [comment]
It's not about performance. It's about having a clean and sane API.
I think `geometry.distance` should always give the `physical` distance, aka minimum image. 
In some cases you might want to cheat / use non physical distances in which case those should have different names.
`mode` might hide them a bit.
But I'm not completely convinced.
Maybe mode is good.
Would like to hear opinions.

**Comment by gcarleo** [comment]
if you want the main problem is where the `type` flag is to be used. I think for the ansatz there is no problem since you will have an ansatz specific to periodic systems that will also know about different types of distance that a `Cell` can accept. Now we have to understand how we make the sampler agnostic somehow of the type of Geometry 

@PhilipVinc the physical distance in a sense is not well defined, it might be that it is L1 that makes sense and not L2, depending on the problem (just to give an example). In the same spirit for the periodic system one type might make more sense than the other type of distance...

**Comment by gpescia** [comment]
> Now we have to understand how we make the sampler agnostic somehow of the type of Geometry

The main difficulty from my view is that for periodic systems you need to fold back positions into the box that went outside, which you don't have to do for free systems. So there is inherently a distinction between the update rules.

**Comment by PhilipVinc** [comment]
But do we need the distance in the sampling?

**Comment by gpescia** [comment]
No in the sampling the distance is irrelevant (right now) except for the model that is sampled.

**Comment by PhilipVinc** [comment]
> The main difficulty from my view is that for periodic systems you need to fold back positions into the box that went outside, which you don't have to do for free systems. So there is inherently a distinction between the update rules.

That's why I was suggesting adding a `geometry.translate` function that takes care of that

**Comment by gcarleo** [comment]
> > Now we have to understand how we make the sampler agnostic somehow of the type of Geometry
> 
> The main difficulty from my view is that for periodic systems you need to fold back positions into the box that went outside, which you don't have to do for free systems. So there is inherently a distinction between the update rules.

True, that's why I was mentioning that implementing vector arithmetic is maybe the clean solution so that the folding is done automatically by the Geometry : x+Delta will fold correctly if the Cell is periodic otherwise it just sums

**Comment by gpescia** [comment]
ah that's a good idea. For the free case this would just return the input and for periodic it would return the backfolded position?

**Comment by PhilipVinc** [comment]
yes

**Comment by gpescia** [comment]
yeah sorry about that.. no idea how tjat commit landed here.
Also I upgraded black and ran it and this resulted in the changes you see.. I‘ll see what is going on.

**Comment by PhilipVinc** [comment]
If you do as above you should get what is specified that is `black==23.9.0`, you can check with `black --version` .
IIRC that thing with the exponentials having extra spacing was from a black a few years back.

**Comment by gpescia** [comment]
Ah seems it falls back on the system version of black rather than the one of the environment when running it as a script.
`python -m black {source_file_or_directory}` works.

**Comment by gpescia** [comment]
Does it destroy back-compatibility if we error out when mixed boundary conditions are used? I have never seen boundary conditions in one direction but not in the other ones in the context of continuous space. The only place this gets used is netket tests. I would be very much in favour for removing this.

**Comment by gcarleo** [comment]
I am not sure it's necessary to keep back compatible interfaces though, this is all in netket.experimental, part of the deal is that we can break these interfaces without major releases

**Comment by PhilipVinc** [comment]
This is not in `netket.experimental`. We could break a bit, but for a good reason (and I'm not sure that creating a geometry is necessary all the time)

**Comment by gpescia** [comment]
I just think that supporting mixed boundary conditions makes it more complicated to take care of all cases while not providing any benefit in terms of usability. If you really want mixed boundary conditions just take pbc everywhere and make one dimension very large... 
But that is up to discussion I guess.

**Comment by PhilipVinc** [comment]
I'm not against dropping support for mixed boundary conditions, though due to MPS it's a not-so-uncommon scenario.
It could raise a not implemented error for now.

**Review by PhilipVinc — COMMENTED** [review]
I would suggest that we agree on the minimal interface of the `ÀbstractGeomtry` object before going forward to avoid wasting time and effort...
The interface means defining every attribute and method name with its signature.
Once this is defined, it can be implemented.

My proposal is

```python
class AbstractGeometry(abc.ABC):
   
   @property
   def n_dimensions(self) -> int:
        # the number of physical dimensions

    def distance(self, x: Array, y: Array) -> Array:
         # computes the physical distance. not necessarily smooth or differentiable.

     def translate(self, x: Array, R : Array) -> Array:
         # translates configuration x by vector R

     def random_uniform(self, shape: Union[int, Tuple[int, ...]], dtype: DType) -> Array:
          # random configurations uniform distribution
```

We still need to: 
 - decide how to compute other distances. Should it be a `mode=...` keyword argument or a different function? if it's a mode, how should it be handled by spaces that don't support it? 
   - Does this mean that we must implement the same `mode` everywhere? What should it be called the `mode` that gives the L2 distance (not minimum image)? 
 - If we decide not to use `mode` how do we write general nn models that work with arbitrary geometries? should we add a `distance_differentiable(x,y)` method to this abstract class?
 - Positioning of this stuff. I propose adding the namespace `netket.experimental.geometry` for the time being and placing this stuff there. Eventually we can move this to `netket.geometry`.
 
@gpescia should confirm if he believes this minimal interface is enough to do everything we need. 

Then, Cell and FreeSpace can add new properties and eventually new methods that are useful. 

What is the `Lattice`argument that Gabriel is using? what does it represent?

**Comment by gcarleo** [comment]
I agree we should first settle on an Abstract Geometry class. I think what @PhilipVinc is ok, but I would make some changes (mostly, I think `translate` is too specific, the general thing is adding vectors or performing other operations) 

```python
class AbstractGeometry(abc.ABC):
   
   @property
   def n_dimensions(self) -> int:
        # the number of physical dimensions

    def distance(self, x: Array, y: Array, mode=None) -> Array:
         # computes the physical distance. not necessarily smooth or differentiable.
         # every geometry implements different modes that are specific of that geometry

    def __add__(self, A: Array, B: Array): -> Array:
         # A+B with folding into pbc if prescribed by the geometry etc 

     def random_uniform(self, shape: Union[int, Tuple[int, ...]], dtype: DType) -> Array:
          # random configurations uniform distribution
```
I also endorse `netket.experimental.geometry` for the time being

File: netket/graph/continuous_geometry.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        """
+        raise NotImplementedError
+
+    def back_to_box(self, x):
+        r"""Given coordinates in space, this function puts the coordinates back to the physical space.
+        For free space this is the identity operation. For periodic boxes the fractional coordinates are
+        put between 0 and 1."""
+        raise NotImplementedError
+
+    def from_lat_to_standard(self, x):
```
[the remark this answers is not in the document]
Comment (gcarleo) [reply netket/graph/continuous_geometry.py:118]: yes this should not be in the base class it's a specific thing of `Cell` I guess

File: netket/graph/continuous_geometry.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    def distance(self, x, y=None, norm=False, mode=None) -> list[list]:
+        """x: (..., N, dim)
+        y: None or (..., N, dim)
+        norm: whether or not the norm of distances is returned as well
+        mode: mode of distance computation
+        """
+        raise NotImplementedError
+
+    def back_to_box(self, x):
```
[the remark this answers is not in the document]
Comment (gcarleo) [reply netket/graph/continuous_geometry.py:112]: `add`, and these methods putting things back should not really be public in my opinion, it's an implementation detail...

**Comment by PhilipVinc** [comment]
`    def __add__(self, A: Array, B: Array): -> Array:` is to implement addition of the object itself, namely geometry, not two third party objects,

File: netket/graph/continuous_geometry.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    def volume(self) -> int:
+        r"""The volume of the Hilbert space if defined (only defined if it is fully periodic)"""
+        raise NotImplementedError
+
+    @property
+    def extent(self):
+        r"""Spatial extension in each spatial dimension"""
+        raise NotImplementedError
+
+    def distance(self, x, y=None, norm=False, mode=None) -> list[list]:
```
[the remark this answers is not in the document]
Comment (gcarleo) [reply netket/graph/continuous_geometry.py:104]: indeed what is norm? `mode` should be enough I guess (e.g. `mode=L2` )

**Comment by gcarleo** [comment]
yes, ok... indeed I think the main issue is that a vector should not be a vector but a `Geometry` object actually

**Comment by gcarleo** [comment]
or, better, a `Geomety` has `Elements` that are what you are calling `Array` now

**Comment by gcarleo** [comment]
or, much easier: 

` def add(self,A,B)` this seems legitimate to me. You just call `geo.add(R1,R2)`

**Comment by PhilipVinc** [comment]
Yes, a `PointNd`, which is generated by the `Geometry` object. 
But let's not get into such details because a custom object would break all of jax and let's implement `Geometry.add(x, y)`, what do you think?

**Comment by PhilipVinc** [comment]
Ok, what about the `mode`argument to distance? what should be the default? what are other possible values we reasonably can think about?

File: netket/graph/continuous_geometry.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gpescia) [reply netket/graph/continuous_geometry.py:104]: If y=None it means distance(x,x) with the caveat that the norm is computed differently in that case.

To not having to call the method twice in case you want the norm and the vectors, there is the extra flag `norm`. We could have a seperate mode keyword for that.

[the remark this answers is not in the document]
Comment (gpescia) [reply netket/graph/continuous_geometry.py:104]: not if x contains N different positions. x can have shape (nsamples, nparticles, sdim) then only the diagonal is 0.

[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/graph/continuous_geometry.py:104]: So in the case  of x being a batch of samples, you return a matrix `nsamples x nsamples` of distances?

Can't we do this by having the following signature for the distance function? 
 - `distance( x:Array[N], y: Array[N]) -> Array[] # (a scalar)`   
 - `distance( x:Array[B1, N], y: Array[N]) -> Array[B1]` 
 - `distance( x:Array[N], y: Array[B2,N]) -> Array[B2]` 
 - `distance( x:Array[B1, N], y: Array[B2,N]) -> Array[B1, B2]` 
 - `distance( x:Array[B1, B2, N], y: Array[B3, N]) -> error`

**Comment by gcarleo** [comment]
It might be good also to check what other establidhed libraries do, I found this for example https://wiki.fysik.dtu.dk/ase/ase/atoms.html#ase.Atoms.get_distance

File: netket/graph/continuous_geometry.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gpescia) [reply netket/graph/continuous_geometry.py:118]: I think it is a naming issue. What lattice is supposed to be is a basis of the physical space. Standard is the standard basis e.g. (1.0),(0,1) in 2D. So a better naming instead of `lattice` is probably `basis`. The two functions then allow to transform back and forth between the two bases.
I actually think it belongs to the base class. I‘m not sure of a use case right now but also in free space it might make sense to define another basis than the standard basis.

**Comment by PhilipVinc** [comment]
@gcarleo one more thing that comes to my mind: geometry could be useful (think necessary) to specify the partition one wants to compute entanglement entropy for in a continuous system. 

It would be useful to think about how to handle/expose specifying a sub-part of the system there.

File: netket/experimental/geometry/continuous_geometry.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        """
+
+    @abc.abstractmethod
+    def add(self, x, y):
+        r"""Given two batches of position vectors in the geometry, this function defines how these vectors are added
+        together. For free space, this is the usual vector addition. FOr periodic space, this includes folding particle
+        positions back to the geometry defined."""
+
+    @abc.abstractmethod
+    def random_init(self, shape):
```
Comment (gcarleo) [inline netket/experimental/geometry/continuous_geometry.py:89]: I wouldn't call this random_init, rather just random_uniform or similar? (the init part is not relevant from the Geometry point of view)

File: netket/experimental/geometry/continuous_geometry.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        r"""This defines an initial configuration in the geometry from which the random_state in Hilbert is constructed.
+        For free space this could just return zeros. For periodic space one could initialize on a lattice.
+        N: int, number of particles."""
+
+    def from_basis_to_standard(self, x):
```
[the remark this answers is not in the document]
Comment (gpescia) [reply netket/experimental/geometry/continuous_geometry.py:94]: For now yes, but I think the defining property of a geometry is a basis, and there might be instances where you want to define a different basis than the standard basis also for free space. Then going back and forth between the custom basis and the standard basis should be part of geometry in my view.

File: netket/hilbert/particle.py [hunk 7]
```diff
@@ -41,52 +43,80 @@ def __init__(
                 if there are no boundaries?). If a finite value is given, a minimum value of zero is assumed for the
                 quantum number(s).
                 A particle in a 3D box of size L would take `(L,L,L)`. A rotor model would take e.g. `(2pi,)`.
-            pbc: Tuple or bool indicating whether to use periodic boundary conditions in a given physical dimension.
-                If tuple it must have the same length as domain. If bool the same value is used for all the dimensions
-                defined in domain.
+            pbc: Tuple of bools or bool indicating whether to use periodic boundary conditions in a given physical
```
Comment (gcarleo) [inline netket/hilbert/particle.py:46]: personally, I would be in favor of breaking the interface instead of adding deprecations and convoluted consutrctors etc... but OK, it's not so user friendly maybe :)

File: netket/experimental/geometry/continuous_geometry.py (hunk 5 above)
Comment (gpescia) [reply netket/experimental/geometry/continuous_geometry.py:89]: For free space I just return zeros, so it is also not really uniform (also not random actually).. So both names are a bit misleading maybe
